[resubmit] Providing more information while crashing process in async error handling

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47246 [resubmit] Providing more information while crashing process in async error handling**

We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

osalpekar committed:
[resubmit] Providing more information while crashing process in async error handling

We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

[ghstack-poisoned]

osalpekar committed:
Update on "[resubmit] Providing more information while crashing process in async error handling"


We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

[ghstack-poisoned]

osalpekar commented:
This change was reverted previous due to failing master-only tests (which is why CI was green) in NCCLErrorHandlingTest. Those tests expected the RuntimeError thrown by NCCL blocking to be "Operation Timed Out!", but this PR changed the message to include more info. This PR updated those tests to expect the correct message.

Previous failed tests here: https://app.circleci.com/pipelines/github/pytorch/pytorch/233465/workflows/19c79d50-b042-4a0d-9765-1efc095a3a29/jobs/8599472

osalpekar committed:
Update on "[resubmit] Providing more information while crashing process in async error handling"


We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -507,12 +513,18 @@ void ProcessGroupNCCL::abortTimedOutCollectives(std::unordered_set<std::string>&
-          << "] Watchdog caught collective operation timeout for work: "
-          << work;
+      auto currentTimepoint = std::chrono::steady_clock::now();
+      auto timeElapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+          currentTimepoint - work.workStartTime_);
+      std::string exceptionMsg = c10::str("[Rank ", rank_, "] ",
+                                          "Watchdog caught collective operation timeout: ",
+                                          work,
+                                          " ran for ",
+                                          timeElapsed.count(),
+                                          " milliseconds before timing out.");
+      LOG(INFO) << exceptionMsg;
       std::exception_ptr exception_ptr = std::make_exception_ptr(
-          std::runtime_error("NCCL Operation Timed Out"));
+          std::runtime_error(exceptionMsg));
```
nit: might be nice to consolidate this into a helper function to avoid duplication above.

osalpekar committed:
Update on "[resubmit] Providing more information while crashing process in async error handling"


We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

[ghstack-poisoned]

osalpekar committed:
Update on "[resubmit] Providing more information while crashing process in async error handling"


We crash the process in NCCL Async Error Handling if the collective
has been running for greater than some set timeout. This PR introduces more
information about the rank and duration the collective ran.

Differential Revision: [D24695126](https://our.internmc.facebook.com/intern/diff/D24695126/)

[ghstack-poisoned]
